lib/integrations/awsoidc/deployservice_iam_config.go

```diff
@@ -0,0 +1,328 @@
+	// Defaults to <TaskRole>Boundary
+	TaskRoleBoundaryPolicyName string
+
+	// AccountID is the AWS Account ID.
+	// Optional. sts.GetCallerIdentity is used if not provided.
+	AccountID string
+
+	// ResourceCreationTags is used to add tags when creating resources in AWS.
+	// Defaults to:
+	// - teleport.dev/cluster: <cluster>
+	// - teleport.dev/origin: aws-oidc-integration
+	// - teleport.dev/integration: <integrationName>
+	ResourceCreationTags awsTags
+
+	// partitionID is the AWS Partition ID.
+	// Eg, aws, aws-cn, aws-us-gov
+	// https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html
+	partitionID string
+}
+
+// CheckAndSetDefaults ensures the required fields are present.
+func (r *DeployServiceIAMConfigureRequest) CheckAndSetDefaults() error {
+	if r.Cluster == "" {
+		return trace.BadParameter("cluster is required")
+	}
+
+	if r.IntegrationName == "" {
+		return trace.BadParameter("integration name is required")
+	}
+
+	if r.Region == "" {
+		return trace.BadParameter("region is required")
+	}
+
+	if r.IntegrationRole == "" {
+		return trace.BadParameter("integration role is required")
+	}
+
+	if r.IntegrationRoleDeployServicePolicy == "" {
+		r.IntegrationRoleDeployServicePolicy = "DeployService"
+	}
+
+	if r.TaskRole == "" {
+		return trace.BadParameter("task role is required")
+	}
+
+	if r.TaskRoleBoundaryPolicyName == "" {
+		r.TaskRoleBoundaryPolicyName = r.TaskRole + boundarySuffix
+	}
+
+	if len(r.ResourceCreationTags) == 0 {
+		r.ResourceCreationTags = DefaultResourceCreationTags(r.Cluster, r.IntegrationName)
+	}
+
+	r.partitionID = awsutils.GetPartitionFromRegion(r.Region)
+
+	return nil
+}
+
+// DeployServiceIAMConfigureClient describes the required methods to create the IAM Roles/Policies required for the DeployService action.
+type DeployServiceIAMConfigureClient interface {
+	// GetCallerIdentity returns information about the caller identity.
+	GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+
+	// CreatePolicy creates a new IAM Policy.
+	CreatePolicy(ctx context.Context, params *iam.CreatePolicyInput, optFns ...func(*iam.Options)) (*iam.CreatePolicyOutput, error)
+
+	// CreateRole creates a new IAM Role.
+	CreateRole(ctx context.Context, params *iam.CreateRoleInput, optFns ...func(*iam.Options)) (*iam.CreateRoleOutput, error)
+
+	// PutRolePolicy creates or replaces a Policy by its name in a IAM Role.
+	PutRolePolicy(ctx context.Context, params *iam.PutRolePolicyInput, optFns ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error)
+}
+
+type defaultDeployServiceIAMConfigureClient struct {
+	*iam.Client
+	stsClient *sts.Client
+}
+
+// NewDeployServiceIAMConfigureClient creates a new DeployServiceIAMConfigureClient.
+func NewDeployServiceIAMConfigureClient(ctx context.Context, region string) (DeployServiceIAMConfigureClient, error) {
```

Comment:
A question that might be outside this PR scope: What about using the [already existing cloud clients](https://github.com/gravitational/teleport/blob/master/lib/cloud/clients.go#L76) for [getting AWS clients](https://github.com/gravitational/teleport/blob/master/lib/cloud/clients.go#L101)? It already initializes the connection (I know the db configurator does not rely on it too, but there is no reason not to use it). The good thing is that AWS already supports some good options, like cross-account clients, and already has a [test implementation](https://github.com/gravitational/teleport/blob/master/lib/cloud/clients.go#L769). That would also help us get everything to use the same SDK major version.

Reply:
My 2c on this, we should really move to v2. I submitted a small request in v1 a few months ago. They rejected and ask me to request it in v2. Maybe we should start a cloud.ClientsV2.

Reply:
Yeah, I really wanted to use V2 here
I could create the same abstraction for V2, but that seemed to be overkill for this PR